docs/topics/compatibility-guides/compatibility-guide-23.md

```diff
@@ -0,0 +1,733 @@
+>
+> **Component**: Gradle
+>
+> **Incompatible change type**: behavioral
+>
+> **Short summary**: In Kotlin 2.3, if you use both the `kotlin-dsl` **and** the `kotlin("jvm")` plugin in your Gradle
+> project, you may see a Gradle warning about an unsupported Kotlin Gradle plugin (KGP) version.
+>
+> **Deprecation cycle**:
+>
+> - 2.3.0: introduce a diagnostic that detects when the `kotlin-dsl` plugin is used with an incompatible language or API version of the compiler.
+> 
+> **Migration steps**:
+> 
+> In general, we don't recommend using both the `kotlin-dsl` and the `kotlin("jvm")` plugins in the same Gradle project. This setup isn't supported.
+> 
+> For convention plugins, precompiled script plugins or any other form of unpublished build logic, you have three options:
+> 
+> 1. Don't apply the `kotlin("jvm")` plugin explicitly. Instead, let the `kotlin-dsl` plugin automatically provide a compatible KGP version.
+> 2. If you want to apply the `kotlin("jvm")` plugin explicitly, use the [`embeddedKotlinVersion`](https://docs.gradle.org/current/kotlin-dsl/gradle/org.gradle.kotlin.dsl/embedded-kotlin-version.html) constant to specify the embedded Kotlin version.
+>
+>     To upgrade the embedded Kotlin and language versions, update your Gradle version. You can find compatible Gradle versions in Gradle's [Compatibility Notes for Kotlin](https://docs.gradle.org/current/userguide/compatibility.html#kotlin).
+> 
+> 3. Don't use the `kotlin-dsl` plugin. This may be more appropriate for binary plugins that aren't tied to a specific Gradle version.
+>
+> As a last resort, you can configure your project to use language version 2.1 or higher, which overrides the conflicting behavior of the `kotlin-dsl` plugin. However, we strongly recommend not doing so.
+> 
+> If you experience difficulties during migration, reach out in the #gradle channel in our [Slack](https://surveys.jetbrains.com/s3/kotlin-slack-sign-up) for support.
+
+### Deprecate `kotlin-android` plugin for AGP versions 9.0.0 and later
+
+> **Issue**: [KT-81199](https://youtrack.jetbrains.com/issue/KT-81199)
+>
+> **Component**: Gradle
+>
+> **Incompatible change type**: source
+>
+> **Short summary**: In Kotlin 2.3.0, the `kotlin-android` plugin is deprecated when using Android Gradle plugin (AGP) versions 9.0.0 or later.
+> Starting with AGP 9.0.0, AGP provides built-in support for Kotlin, so the `kotlin-android` plugin is no longer needed.
+>
+> **Deprecation cycle**:
+>
+> - 2.3.0: report a warning when the `kotlin-android` plugin is used with AGP versions 9.0.0 or later, and both the `android.builtInKotlin` and `android.newDsl=false` Gradle properties are set to `false`.
+
+### Deprecate `testApi` configuration
+
+> **Issue**: [KT-63285](https://youtrack.jetbrains.com/issue/KT-63285)
+>
+> **Component**: Gradle
+>
+> **Incompatible change type**: source
+>
+> **Short summary**: Kotlin 2.3.0 deprecates the `testApi` configuration since it's an unnecessary addition to Gradle projects.
```

Comment:
I would rephrase it. `testApi` configuration means test sources could be consumed in some other project. But it is not possible in Gradle and for sharing test code Gradle has a special feature - [text fixtures](https://docs.gradle.org/current/userguide/java_testing.html#sec:java_test_fixtures)